### PR TITLE
Introduce WindowInsets

### PR DESCRIPTION
**Pull request by noahwe145**
## Purpose / Description
Deprecated UI-Flags should be replaced by WindowInsets.  

## Fixes
Fixes https://github.com/ankidroid/Anki-Android/issues/9332

## Approach
WindowInsets and WindowInsetsController are used

## How Has This Been Tested?
Unit tests
979 tests completed, 560 failed, 17 skipped

_Links to blog posts, patterns, libraries or addons used to solve this problem_
https://stackoverflow.com/questions/62643517/immersive-fullscreen-on-android-11
https://stackoverflow.com/questions/64151280/android-r-api-30-system-bar-visibility-change-call-backs

## Checklist
_Please, go through these checks before submitting the PR._

- [x] You have not changed whitespace unnecessarily (it makes diffs hard to read)
- [x] You have a descriptive commit message with a short title (first line, max 50 chars).
- [x] Your code follows the style of the project (e.g. never omit braces in `if` statements) 
- [x] You have commented your code, particularly in hard-to-understand areas
- [x] You have performed a self-review of your own code
- [x] UI changes: include screenshots of all affected screens (in particular showing any new or changed strings)
- [x] UI Changes: You have tested your change using the [Google Accessibility Scanner](https://play.google.com/store/apps/details?id=com.google.android.apps.accessibility.auditor)

### CODE REVIEW

**Review by david-allison-1 — CHANGES_REQUESTED** [review]
Thanks for the PR!

This adds `@RequiresApi(api = Build.VERSION_CODES.R)` to many vital methods, we should instead be using, or adding methods to our `Compat`/`CompatHandler` infrastructure:

https://github.com/ankidroid/Anki-Android/tree/0c1a38ea34babb569ff5c1ff17ab96b337a55bb8/AnkiDroid/src/main/java/com/ichi2/compat

**Comment by noahwe145** [comment]
i added
`if (Build.VERSION.SDK_INT >= Build.VERSION_CODES.R) {
       changes
} else {
         deprecated code
 }
`
to all changes so it should be backward compatible

File: AnkiDroid/src/main/java/com/ichi2/anki/AnkiActivity.java [hunk 1]
```diff
@@ -88,8 +91,14 @@ protected void onCreate(Bundle savedInstanceState) {
...
-        if (AdaptionUtil.isUserATestClient()) {
-            getWindow().setFlags(WindowManager.LayoutParams.FLAG_FULLSCREEN, WindowManager.LayoutParams.FLAG_FULLSCREEN);
+        if (Build.VERSION.SDK_INT >= Build.VERSION_CODES.R) {        if (AdaptionUtil.isUserATestClient()) {
+            getWindow().getDecorView().getWindowInsetsController().hide(android.view.WindowInsets.Type.statusBars());
+        }
+        else {
+            if (AdaptionUtil.isUserATestClient()) {
+                getWindow().setFlags(WindowManager.LayoutParams.FLAG_FULLSCREEN, WindowManager.LayoutParams.FLAG_FULLSCREEN);
+            }
+        }
```
Comment (david-allison-1) [inline AnkiDroid/src/main/java/com/ichi2/anki/AnkiActivity.java:101]: Thanks! Could you move this into our `compat` infrastructure.

See the following for an example:

https://github.com/ankidroid/Anki-Android/tree/0c1a38ea34babb569ff5c1ff17ab96b337a55bb8/AnkiDroid/src/main/java/com/ichi2/compat

You can also Ctrl+F for `CompatHelper.getCompat()` to find usages

**Review by david-allison-1 — CHANGES_REQUESTED** [review]

File: AnkiDroid/src/main/java/com/ichi2/anki/AnkiActivity.java [hunk 2]
```diff
@@ -91,14 +92,10 @@ protected void onCreate(Bundle savedInstanceState) {
...
-        }
-        else {
-            if (AdaptionUtil.isUserATestClient()) {
-                getWindow().setFlags(WindowManager.LayoutParams.FLAG_FULLSCREEN, WindowManager.LayoutParams.FLAG_FULLSCREEN);
-            }
-        }
+
+        if (AdaptionUtil.isUserATestClient()) {
+            Compat compat = CompatHelper.getCompat();
+            compat.setFullscreen(getWindow());
```
Comment (david-allison-1) [inline AnkiDroid/src/main/java/com/ichi2/anki/AnkiActivity.java:98]: Much nicer! Just a nitpick

```suggestion
            CompatHelper.getCompat().setFullscreen(getWindow();
```

File: AnkiDroid/src/main/java/com/ichi2/anki/AnkiActivity.java [hunk 3]
```diff
@@ -88,8 +92,10 @@ protected void onCreate(Bundle savedInstanceState) {
         Themes.setTheme(this);
         super.onCreate(savedInstanceState);
         // Disable the notifications bar if running under the test monkey.
+
```
Comment (david-allison-1) [inline AnkiDroid/src/main/java/com/ichi2/anki/AnkiActivity.java:95]: nitpick: could you avoid changing whitespace in a diff

File: AnkiDroid/src/main/java/com/ichi2/anki/Reviewer.java [hunk 4]
```diff
@@ -1155,47 +1157,82 @@ private void setWhiteboardEnabledState(boolean state) {
...
-                View.SYSTEM_UI_FLAG_LAYOUT_STABLE
-                        //| View.SYSTEM_UI_FLAG_LAYOUT_HIDE_NAVIGATION // temporarily disabled due to #5245
-                        | View.SYSTEM_UI_FLAG_LAYOUT_FULLSCREEN
-                        | View.SYSTEM_UI_FLAG_HIDE_NAVIGATION
-                        | View.SYSTEM_UI_FLAG_FULLSCREEN
-                        | View.SYSTEM_UI_FLAG_LOW_PROFILE
-                        | View.SYSTEM_UI_FLAG_IMMERSIVE
-        );
+        //Hide System Bars
+        if (Build.VERSION.SDK_INT >= Build.VERSION_CODES.R) {
```
Comment (david-allison-1) [inline AnkiDroid/src/main/java/com/ichi2/anki/Reviewer.java:1161]: Could you go through and convert each of these usages to the Compat infrastructure?

**Review by david-allison-1 — CHANGES_REQUESTED** [review]
Thanks! Looks good. Could you do the same (using `Compat`) for the other changes

File: AnkiDroid/src/main/java/com/ichi2/anki/Reviewer.java (hunk 4 above)
Comment (noahwe145) [reply AnkiDroid/src/main/java/com/ichi2/anki/Reviewer.java:1161]: yeah sure

**Comment by noahwe145** [comment]
i moved all changes i could, some were not possible because they use private functions of Reviewer

**Comment by david-allison-1** [comment]
Thanks, please follow up if I haven't re-reviewed this by Sunday

**Comment by noahwe145** [comment]
@david-allison-1  sorry, forgot to remind you

File: AnkiDroid/src/main/java/com/ichi2/anki/AnkiActivity.java [hunk 5]
```diff
@@ -12,12 +12,14 @@
 import android.graphics.Color;
 import android.media.AudioManager;
 import android.net.Uri;
+import android.os.Build;
```
Comment (david-allison-1) [inline AnkiDroid/src/main/java/com/ichi2/anki/AnkiActivity.java:15]: (minor/nitpick) a few of these imports are no longer needed

File: AnkiDroid/src/main/java/com/ichi2/anki/AnkiActivity.java [hunk 6]
```diff
@@ -96,7 +100,7 @@ protected void onCreate(Bundle savedInstanceState) {
         super.onCreate(savedInstanceState);
         // Disable the notifications bar if running under the test monkey.
         if (AdaptionUtil.isUserATestClient()) {
-            getWindow().setFlags(WindowManager.LayoutParams.FLAG_FULLSCREEN, WindowManager.LayoutParams.FLAG_FULLSCREEN);
+            CompatHelper.getCompat().setFullscreen(getWindow());
```
Comment (david-allison-1) [inline AnkiDroid/src/main/java/com/ichi2/anki/AnkiActivity.java:103]: Looks great!

File: AnkiDroid/src/main/java/com/ichi2/anki/Reviewer.java [hunk 7]
```diff
@@ -1162,49 +1165,66 @@ private void setWhiteboardEnabledState(boolean state) {
...
-                View.SYSTEM_UI_FLAG_LAYOUT_STABLE
-                        //| View.SYSTEM_UI_FLAG_LAYOUT_HIDE_NAVIGATION // temporarily disabled due to #5245
-                        | View.SYSTEM_UI_FLAG_LAYOUT_FULLSCREEN
-                        | View.SYSTEM_UI_FLAG_HIDE_NAVIGATION
-                        | View.SYSTEM_UI_FLAG_FULLSCREEN
-                        | View.SYSTEM_UI_FLAG_LOW_PROFILE
-                        | View.SYSTEM_UI_FLAG_IMMERSIVE
-        );
+        //Hide System Bars
+        CompatHelper.getCompat().hideSystembars(getWindow());
```
Comment (david-allison-1) [inline AnkiDroid/src/main/java/com/ichi2/anki/Reviewer.java:1171]: nit: Could you rename the method to `hideSystemBars`

File: AnkiDroid/src/main/java/com/ichi2/anki/Reviewer.java [hunk 8]
```diff
@@ -1162,49 +1165,66 @@ private void setWhiteboardEnabledState(boolean state) {
...
-        a.getWindow().getDecorView().setSystemUiVisibility(
-                View.SYSTEM_UI_FLAG_LAYOUT_STABLE
-                        //| View.SYSTEM_UI_FLAG_LAYOUT_HIDE_NAVIGATION // temporarily disabled due to #5245
-                        | View.SYSTEM_UI_FLAG_LAYOUT_FULLSCREEN
-                        | View.SYSTEM_UI_FLAG_HIDE_NAVIGATION
-                        | View.SYSTEM_UI_FLAG_FULLSCREEN
-                        | View.SYSTEM_UI_FLAG_LOW_PROFILE
-                        | View.SYSTEM_UI_FLAG_IMMERSIVE
-        );
+        //Hide System Bars
```
Comment (david-allison-1) [inline AnkiDroid/src/main/java/com/ichi2/anki/Reviewer.java:1170]: Doesn't need a comment - the method name is good enough

```suggestion
```

File: AnkiDroid/src/main/java/com/ichi2/anki/VideoPlayer.java [hunk 9]
```diff
@@ -42,8 +45,9 @@ protected void onCreate(Bundle savedInstanceState) {
         setContentView(R.layout.video_player);
         mPath = getIntent().getStringExtra("path");
         Timber.i("Video Player intent had path: %s", mPath);
-        getWindow().setFlags(WindowManager.LayoutParams.FLAG_FULLSCREEN,
-                WindowManager.LayoutParams.FLAG_FULLSCREEN);        
+
+        CompatHelper.getCompat().setFullscreen(getWindow());
+
```
Comment (david-allison-1) [inline AnkiDroid/src/main/java/com/ichi2/anki/VideoPlayer.java:50]:
```suggestion
        CompatHelper.getCompat().setFullscreen(getWindow());
```

File: AnkiDroid/src/main/java/com/ichi2/compat/Compat.java [hunk 10]
```diff
@@ -83,6 +84,7 @@
     long copyFile(String source, OutputStream target) throws IOException;
     long copyFile(InputStream source, String target) throws IOException;
 
+
```
Comment (david-allison-1) [inline AnkiDroid/src/main/java/com/ichi2/compat/Compat.java:87]: Aim to not change whitespace - it makes the diffs/git history harder to reason about 

```suggestion
```

File: AnkiDroid/src/main/java/com/ichi2/compat/CompatV30.java [hunk 11]
```diff
@@ -0,0 +1,40 @@
+
```
Comment (david-allison-1) [inline AnkiDroid/src/main/java/com/ichi2/compat/CompatV30.java:1]: This needs a copyright header. There should be a lint warning here with instructions on how to set this to be inserted manually

File: AnkiDroid/src/main/java/com/ichi2/compat/CompatV30.java [hunk 12]
```diff
@@ -0,0 +1,40 @@
...
+import android.view.View;
+import android.view.Window;
+import android.view.WindowInsets;
+import android.view.WindowInsetsController;
+
+import com.ichi2.anki.R;
+import com.ichi2.anki.reviewer.FullScreenMode;
+
+
+/** Implementation of {@link Compat} for SDK level 30 and higher. Check  {@link Compat}'s for more detail. */
```
Comment (david-allison-1) [inline AnkiDroid/src/main/java/com/ichi2/compat/CompatV30.java:18]: I love this comment, great!

```suggestion
/** Implementation of {@link Compat} for SDK level 30 and higher. Check {@link Compat}'s for more detail. */
```

File: AnkiDroid/src/main/java/com/ichi2/compat/CompatV30.java [hunk 13]
```diff
@@ -0,0 +1,40 @@
...
+@TargetApi(30)
+public class CompatV30 extends CompatV26 implements Compat {
+    @Override
+    public void setFullscreen(Window window) {
+        window.getDecorView().getWindowInsetsController().hide(WindowInsets.Type.statusBars());
```
Comment (david-allison-1) [inline AnkiDroid/src/main/java/com/ichi2/compat/CompatV30.java:23]: Could you add a comment to the interface listing what this does.

I'd expect the soft buttons (home/apps/back) to be hidden as well. Does this happen?

File: AnkiDroid/src/main/java/com/ichi2/compat/CompatV30.java [hunk 14]
```diff
@@ -0,0 +1,40 @@
...
+    }
+
+    @Override
+    public void hideSystembars(Window window) {
+        window.setDecorFitsSystemWindows(false);
```
Comment (david-allison-1) [inline AnkiDroid/src/main/java/com/ichi2/compat/CompatV30.java:28]: Could you add a reference for where this code came from?

File: AnkiDroid/src/main/java/com/ichi2/compat/CompatV30.java [hunk 15]
```diff
@@ -0,0 +1,40 @@
...
+        WindowInsetsController controller = window.getInsetsController();
+        if (controller != null) {
+            controller.hide(WindowInsets.Type.statusBars() | WindowInsets.Type.navigationBars());
+            controller.setSystemBarsBehavior(WindowInsetsController.BEHAVIOR_SHOW_TRANSIENT_BARS_BY_SWIPE);
+        }
+    }
+
+    @Override
+    public boolean isImmersiveSystemUiVisible(Window window) {
+        return (window.getDecorView().getWindowInsetsController().getSystemBarsAppearance()) == 0;
```
Comment (david-allison-1) [inline AnkiDroid/src/main/java/com/ichi2/compat/CompatV30.java:38]: There's no need for the brackets here.

Could you add a reference for why this is the case? 

0 maps to the deprecated https://developer.android.com/reference/android/view/WindowInsetsController#BEHAVIOR_SHOW_BARS_BY_TOUCH

And I haven't checked whether `WindowInsets.Type.statusBars() | WindowInsets.Type.navigationBars()` would map to 0 in all cases

File: AnkiDroid/src/main/java/com/ichi2/compat/Compat.java [hunk 16]
```diff
@@ -144,6 +146,10 @@
     void requestAudioFocus(AudioManager audioManager, AudioManager.OnAudioFocusChangeListener audioFocusChangeListener, @Nullable AudioFocusRequest audioFocusRequest);
     void abandonAudioFocus(AudioManager audioManager, AudioManager.OnAudioFocusChangeListener audioFocusChangeListener, @Nullable AudioFocusRequest audioFocusRequest);
 
+    void setFullscreen(Window window);
```
Comment (david-allison-1) [inline AnkiDroid/src/main/java/com/ichi2/compat/Compat.java:149]: These would be great with comments to ensure that we've handled all the functionality

**Review by david-allison-1 — COMMENTED** [review]
Sorry about the delay! This wasn't a good one to review on the phone.

A few nitpicks, my main concern is adding documentation as this is a fairly new area.

My other concern is `isImmersiveSystemUiVisible`. 

I'll further test this on my Android 11 after the stylistic changes and documentation are added

**Comment by david-allison-1** [comment]
@noahwe145 is this still active? (no worries if it's not)

**Comment by mikehardy** [comment]
This appears to need a new dev, a lot of work has been invested in it already and it appears close to working to me if someone wants to pick it up, resolve the current conflicts and get it polished up for merge

**Comment by mikehardy** [comment]
Has current work in #11531

**Pull request closed without merging**